### ISSUE TITLE
[BUG]: Transformers getting pre-transformed data.

### ISSUE THREAD

**Issue by spearmootz** [sensedeep/dynamodb-onetable#111]
transform function should be split into getters and setters in order to be able to add types to something.

so for example, lets say we have type Object but i want to know why kind of object it is when i am dealing with it then it would get the get return type.

or lets say that i want to store this object as a string instead and retrieve it back as an object i could specify type, get and set functions.

**Comment by mobsense** [comment]
I don't fully understand your suggestion.  When you say add "types" what do you mean?  Do you mean add extra attributes/properties or do you mean map the type of an attribute value.

There are 3 possible transform hooks. See below. Two work on the entire record. One works on field by field.
In all, you can map the actual value or its type to and from the database.

The schema transform is:

            value = field.transform(this, 'write', field.name, value)

I can't seem to find the params.transform in the doc. The signature is:

            rec = params.transform(this, 'write', rec, params)

You can also apply a table level hook via the intercept Table construct param.

            rec = this.table.intercept(this, op, rec, params)

**Comment by spearmootz** [comment]
these in some way do not serve the purpose.  so let me illustrate in some way Dynamo-Onetable accepts Date as a type, but the underlying data type is a string. so there is convenience to having the storage data type be different than the one utilized in the application.

perhaps an example will illustrate the point.  The following should work but it doesnt for multiple reasons, the value is validated to be a string before it is transformed and if it did Typescript would not know whether the value accessed is a string or ParseOptions because they are intertwined in one function.

parseOptions: {
    type: String,
    required: true,
    map: 'pa',
    transform: (
      _model: any,
      op: string,
      _name: any,
      parseOptions: string | ParseOptions
    ): string | ParseOptions => {
      console.log('TRANSFORM');

      return op == 'read'
        ? JSON.parse(parseOptions as string)
        : JSON.stringify(parseOptions as ParseOptions);
    },
  },

**Comment by spearmootz** [comment]
here is what i am saying it should be like.

```
parseOptions: {
    type: String,
    required: true,
    map: 'pa',
    get: (  _model: any,
      op: string,
      _name: any,
      parseOptions: string): ParseOptions => JSON.parse(parseOptions)
      set: (  _model: any,
      op: string,
      _name: any,
      parseOptions: ParseOptions): string => JSON.stringify(parseOptions)
  },
  ```

**Comment by mobsense** [comment]
Thanks. That helps.

One question. You seem to be indicating that the 4th parameter 'params' is of a different type for read vs write. Isn't the 4th parameter always just the value?  I'm not sure I understand why you have JSON.parse(parseOptions as string) vs parse(parseOptions as ParseOptions) (what is ParseOptions?).

I'd like to capture the underlying issues and then consider solutions.  While your solution certainly has merit, I want to really understand all of the underling issues to make sure we are addressing them all. As we already have 3 transform options, I'd really like to rationalize them and get a unified, more flexible solution that meets all the needs.

Looks like the issues are:

* The field.transform for write is invoked AFTER that value is already transformed by the inbuilt type transformers. This happens in Model.transformWriteAttribute
* The field.transform for read is invoked ONLY and bypasses the inbuilt type transformers in Model.transformReadAttribute.  The schema transformer should be consistent with read vs write.
* The field.transform is not receiving the `field` information which contains the schema field type,
* The params.transform gets called AFTER the field.transform. i.e. both get called whereas params.transform should take precedence.

Are there others I've missed?

**Comment by spearmootz** [comment]
The answer to your question is, the get should expect the underlying data type and transform it to the desired and the set should receive the desired data type and transform it to the underlying data type.

ParseOptions in that case would be the type of object i want to work with. In the case of a date it could me a moment instance instead of the normal date object. 

It could be that i want to save a string but in reality it represents a RegEx, so I set the value to a RegEx, the transform executes which takes it from a RegEx to a string and when i retrieve it back from the DB the string gets turned into a RegEx.

**Comment by mobsense** [comment]
I understand that and agree that is the purpose of the transform function. To allow users to do exactly that.

The reason you are not getting the stored data as you wish is the schema transformer is conflicting with the inbuilt transformers. If you define a transformer, it should be invoked ONLY and bypass the inbuilt transformers.

So the signature for the "value" parameter on read would be ANY as you could store the data type anyway you wish. Similarly the return value could be any type. You can transform any to any.  

Given that, I would have thought that read vs write would be symmetrical. Why are two functions required (get & set) ?

**Comment by spearmootz** [comment]
For Typescript purposes ,i thought it would be nice to have it nicely typed.

**Comment by mobsense** [comment]
But if the value can be any and the return value can be any?

**Comment by spearmootz** [comment]
Right, seeing l some inference would need to happen to set the model type correctly

**Comment by mobsense** [comment]
I'll change the designation to BUG and we'll fix the clashing with the inbuilt transformers. Then could you see if that will work as you need. I'm happy to revisit after we fix the core bug where your transformer gets pre-transformed data.

**Comment by mobsense** [comment]
Fixed in 1.8.1. The schema field.transform will be invoked without any of the inbuilt transformations.
